### PR TITLE
Temp sensing

### PR DESCRIPTION
**Pull request by zx100x100**
Add support for a temperature sensing message. tailored for cansw_temperature_sensing, but designed to be generic enough for any future use with other things.

Check for yourself, but all unit tests pass

---
This change is [Reviewable](https://reviewable.io/reviews/waterloo-rocketry/canlib/36)

### CODE REVIEW

**Review by BluCodeGH — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/canlib/36)* status: all files reviewed, 6 unresolved discussions (waiting on @BenLevy7, @jacobdeery, @JaredWatson, @LanaTomlin, @manumanuk, @Rio6, @sineadazam, @wendi-yu, @ZTeper, and @zx100x100)

___
*[can_common.h, line 146 at r1](https://reviewable.io/reviews/waterloo-rocketry/canlib/36#-MfApPtH9gumU1KhWgt8:-MfApPtI7ohuBl4VJWon:b-312yc6) ([raw file](https://github.com/waterloo-rocketry/canlib/blob/707f0daf9d2b70a25e86095138155beadee6961f/can_common.h#L146)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> bool build_temp_data_msg(uint32_t timestamp,
>                              uint8_t sensor_num,
>                              int32_t temp,
>                              can_msg_t *output);
> ```
</details>

Indentation here is wrong
___
*[can_common.h, line 306 at r1](https://reviewable.io/reviews/waterloo-rocketry/canlib/36#-MfApUy21ahSQMm4-xF2:-MfApUy21ahSQMm4-xF3:b-a7wp74) ([raw file](https://github.com/waterloo-rocketry/canlib/blob/707f0daf9d2b70a25e86095138155beadee6961f/can_common.h#L306)):*
> ```
> bool get_temp_data(const can_msg_t *msg,
>                        uint8_t *sensor_num,
>                        int32_t *temp);
> ```
Indentation
___
*[can_common.c, line 256 at r1](https://reviewable.io/reviews/waterloo-rocketry/canlib/36#-MfApbFL6ToX2YwLqjc2:-MfApbFL6ToX2YwLqjc3:b-p6aadt) ([raw file](https://github.com/waterloo-rocketry/canlib/blob/707f0daf9d2b70a25e86095138155beadee6961f/can_common.c#L256)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> bool build_temp_data_msg(uint32_t timestamp,
>                              uint8_t sensor_num,
>                              int32_t temp,
>                              can_msg_t *output){
> ```
</details>

Indentation and opening bracket placement.
___
*[can_common.c, line 259 at r1](https://reviewable.io/reviews/waterloo-rocketry/canlib/36#-MfAppHO13YdggA3i92S:-MfAppHO13YdggA3i92T:by4h4ee) ([raw file](https://github.com/waterloo-rocketry/canlib/blob/707f0daf9d2b70a25e86095138155beadee6961f/can_common.c#L259)):*
> ```C
>     if(!output) { return false; }
> 
>     output->sid =  MSG_SENSOR_TEMP | BOARD_UNIQUE_ID;
> ```

Double space after the =
___
*[can_common.c, line 639 at r1](https://reviewable.io/reviews/waterloo-rocketry/canlib/36#-MfAq3Wo43_DaLGFAEG1:-MfAq3Wo43_DaLGFAEG2:bes8s1d) ([raw file](https://github.com/waterloo-rocketry/canlib/blob/707f0daf9d2b70a25e86095138155beadee6961f/can_common.c#L639)):*
> ```
> bool get_temp_data(const can_msg_t *msg,
>                        uint8_t *sensor_num,
>                        int32_t *temp)
> ```
Indentation
___
*[tests/unit/build_can_message.c, line 445 at r1](https://reviewable.io/reviews/waterloo-rocketry/canlib/36#-MfAqq99DrsfKLO_FcLk:-MfAqq99DrsfKLO_FcLl:biwyyi4) ([raw file](https://github.com/waterloo-rocketry/canlib/blob/707f0daf9d2b70a25e86095138155beadee6961f/tests/unit/build_can_message.c#L445)):*
> ```C
> {
>     uint8_t sensor_num = 27;
>     int32_t temp = -3000*1024;
> ```

Would be good to test both positive and negative temperatures to verify the sign conversion.

File: can_common.c [hunk 1]
```diff
@@ -250,6 +250,24 @@ bool build_analog_data_msg(uint32_t timestamp,
     return true;
 }
 
+bool build_temp_data_msg(uint32_t timestamp,
+                             uint8_t sensor_num,
+                             int32_t temp,
```
Comment (Rio6) [inline can_common.c:255]: Should we add a comment saying temp is 24 bit singned integer? (and numbers >= 2^23 would become negative).

**Review by wendi-yu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/canlib/36)* status: all files reviewed, 8 unresolved discussions (waiting on @BenLevy7, @jacobdeery, @JaredWatson, @LanaTomlin, @manumanuk, @sineadazam, @ZTeper, and @zx100x100)

___
*[tests/unit/build_can_message.c, line 461 at r1](https://reviewable.io/reviews/waterloo-rocketry/canlib/36#-Mfo2pzU3nMsq2z7h4nX:-Mfo2pzVAE6bdlDAak-1:bl2oekc) ([raw file](https://github.com/waterloo-rocketry/canlib/blob/707f0daf9d2b70a25e86095138155beadee6961f/tests/unit/build_can_message.c#L461)):*
> ```C
>         ret = false;
>     }
>     if (output.sid != 0x543) {
> ```

Maybe worth pulling the `0x543` out into a named variable?

**Comment by zx100x100** [comment]
> > _[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/canlib/36)_ status: all files reviewed, 8 unresolved discussions (waiting on @BenLevy7, @jacobdeery, @JaredWatson, @LanaTomlin, @manumanuk, @sineadazam, @ZTeper, and @zx100x100)
> 
> _[tests/unit/build_can_message.c, line 461 at r1](https://reviewable.io/reviews/waterloo-rocketry/canlib/36#-Mfo2pzU3nMsq2z7h4nX:-Mfo2pzVAE6bdlDAak-1:bl2oekc) ([raw file](https://github.com/waterloo-rocketry/canlib/blob/707f0daf9d2b70a25e86095138155beadee6961f/tests/unit/build_can_message.c#L461)):_
> 
> > ```c
> >         ret = false;
> >     }
> >     if (output.sid != 0x543) {
> > ```
> 
> Maybe worth pulling the `0x543` out into a named variable?

Might be nice to re-factor the unit tests to do this at some point, but this is how all the tests do it, so best to keep it this way for consistency for now.

File: can_common.c (hunk 1 above)
Comment (zx100x100) [reply can_common.c:255]: placed in header file, but done

**Review by BluCodeGH — DISMISSED** [review]
:lgtm:
I'd like to hold off on merging this so we can get #37 in, since it changes the column spacing in the comment in message_types.h and it will be a lot easier to update one line here than for me to re-fix merge conflicts.

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/canlib/36#-:-Ms196i5C7n_B1lKJvuF:b-azfbjx)* status: all files reviewed, 2 unresolved discussions (waiting on @BenLevy7, @JaredWatson, @LanaTomlin, @manumanuk, @Rio6, @sineadazam, @wendi-yu, and @ZTeper)

**Review by BluCodeGH — APPROVED** [review]
Merge conflicts fixed

**Pull request merged**